### PR TITLE
feat: remove typechain

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This PR:

 - removes typechain in favor of viem contracts and saner types
 - fixes the rugged ERC20ABI, cleans up the ABI files, and converts them to JSON with an exported `as const` object for type inference
 - fixes some viem usages in investor-idle

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A

## Risk

Moderate.

- THORChain savers deposit/withdraws/approve for EVM assets may be broken
- ETH/FOX LP deposit/withdraw as well as APR may be broken
- FOX farming deposit/withdraw as well as APR may be broken

## Testing

- See risks. THOR savers and ETH/FOX LP (and other UNI-V2 pools by extension)/staking may be broken, test accordingly 

### Engineering

- Refactor to viem looks sane

### Operations

- ☝🏽 

### Balances and APY

- Develop

Screenshot 2023-10-25 at 20 00 01
Screenshot 2023-10-25 at 20 00 13

- This diff

Screenshot 2023-10-25 at 20 01 28
Screenshot 2023-10-25 at 20 01 41

### Withdraw

image

### Deposit

image

### CODE REVIEW

**Comment by gomesalexandre** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #5532** <a href="https://app.graphite.dev/github/pr/shapeshift/web/5532" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/5532?utm_source=stack-comment).

File: src/contracts/abis/ERC20ABI.json
Comment (gomesalexandre) [inline src/contracts/abis/ERC20ABI.json:1]: I lost way too much sanity on this:
1. our ABI was rugged and made all viem type inferences broken
2. wagmi exposes a sane version of the ABI so we should keep it DRY October

Comment (gomesalexandre) [reply src/contracts/abis/ERC20ABI.json:1]: Disregard 2 - importing for wagmi turbo rugs tests - ee638166af70bbc668733f205a460316fac9381b

File: src/state/slices/opportunitiesSlice/resolvers/uniV2/utils.test.ts [hunk 1]
```diff
@@ -7,76 +7,55 @@ import BigNumber from 'bignumber.js'
...
+}))
 
 const mockContract = getContract({
   abi: [],
   address: '' as Address,
 }) as unknown as GetContractReturnType<typeof IUniswapV2Pair, PublicClient, WalletClient>
 const token0Decimals = 18
-const mockToken0Reserves = new TokenAmount(new Token(1, '', token0Decimals), tokenAmount)
+const mockToken0Reserves = new TokenAmount(
+  new Token(1, '0x0000000000000000000000000000000000000000', token0Decimals),
```
Comment (gomesalexandre) [inline src/state/slices/opportunitiesSlice/resolvers/uniV2/utils.test.ts:51]: Looks like this was never actually ran at runtime - this throws and expects a valid address

File: src/state/slices/opportunitiesSlice/resolvers/uniV2/utils.test.ts [hunk 2]
```diff
@@ -93,13 +72,13 @@ describe('resolvers/univ2/utils', () => {
...
-    const totalDailyVolume = new BigNumber(amount0In).plus(
-      new BigNumber(amount0Out).div(1 - TRADING_FEE_RATE).decimalPlaces(0),
+    const totalDailyVolume = new BigNumber(mockAmount0In).plus(
+      new BigNumber(mockAmount0Out).div(1 - TRADING_FEE_RATE).decimalPlaces(0),
     )
 
     const totalAnnualVolume = totalDailyVolume.times(365.25)
     const totalAnnualRewards = totalAnnualVolume.times(TRADING_FEE_RATE).decimalPlaces(0)
-    const token0ReservesBaseUnitValue = new BigNumber(tokenAmount).times(`1e+${token0Decimals}`)
+    const token0ReservesBaseUnitValue = new BigNumber(tokenAmount)
```
Comment (gomesalexandre) [inline src/state/slices/opportunitiesSlice/resolvers/uniV2/utils.test.ts:81]: Looks weird but is correct - tests were previously off, confirmed at runtime we're still 🛍️

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts [hunk 3]
```diff
@@ -173,9 +177,12 @@ export const useFoxFarming = (
     (lpAmount: string) => {
       if (skip || !adapter || !isValidAccountNumber(accountNumber) || !wallet) return
 
-      const data = foxFarmingContract.interface.encodeFunctionData('stake', [
-        toBaseUnit(lpAmount, lpAsset.precision),
-      ])
+      const data = encodeFunctionData({
+        // @ts-ignore this is drunk, the ABI is isValid
```
Comment (gomesalexandre) [inline src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts:181]: revert

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts [hunk 4]
```diff
@@ -193,11 +200,12 @@ export const useFoxFarming = (
...
 
-      const data = isExiting
-        ? foxFarmingContract.interface.encodeFunctionData('exit')
-        : foxFarmingContract.interface.encodeFunctionData('withdraw', [
-            toBaseUnit(lpAmount, lpAsset.precision),
-          ])
+      const data = encodeFunctionData({
+        // @ts-ignore this is drunk, the ABI is isVali
+        abi: foxFarmingContract.abi,
+        function: isExiting ? 'exit' : 'withdraw',
```
Comment (gomesalexandre) [inline src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts:206]:
```suggestion
        functionName: isExiting ? 'exit' : 'withdraw',
```

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts [hunk 5]
```diff
@@ -193,11 +200,12 @@ export const useFoxFarming = (
     (lpAmount: string, isExiting: boolean) => {
       if (skip || !adapter || !isValidAccountNumber(accountNumber) || !wallet) return
 
-      const data = isExiting
-        ? foxFarmingContract.interface.encodeFunctionData('exit')
-        : foxFarmingContract.interface.encodeFunctionData('withdraw', [
-            toBaseUnit(lpAmount, lpAsset.precision),
-          ])
+      const data = encodeFunctionData({
+        // @ts-ignore this is drunk, the ABI is isVali
```
Comment (gomesalexandre) [inline src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts:204]: revert

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts [hunk 6]
```diff
@@ -215,7 +223,11 @@ export const useFoxFarming = (
     async (userAddress: string) => {
       if (!adapter || !userAddress || !wallet) return
 
-      const data = foxFarmingContract.interface.encodeFunctionData('getReward')
+      const data = encodeFunctionData({
+        // @ts-ignore this is drunk, the ABI is isValid
+        abi: foxFarmingContract.abi,
+        function: 'getReward',
```
Comment (gomesalexandre) [inline src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts:229]:
```suggestion
        functionName: 'getReward',
```

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts [hunk 7]
```diff
@@ -215,7 +223,11 @@ export const useFoxFarming = (
     async (userAddress: string) => {
       if (!adapter || !userAddress || !wallet) return
 
-      const data = foxFarmingContract.interface.encodeFunctionData('getReward')
+      const data = encodeFunctionData({
+        // @ts-ignore this is drunk, the ABI is isValid
```
Comment (gomesalexandre) [inline src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts:227]: revert

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts [hunk 8]
```diff
@@ -234,10 +246,12 @@ export const useFoxFarming = (
...
     if (!adapter) throw new Error(`no adapter available for ${ethAsset.chainId}`)
 
-    const data = uniV2LPContract.interface.encodeFunctionData('approve', [
-      contractAddress,
-      MaxUint256,
-    ])
+    const data = encodeFunctionData({
+      // @ts-ignore this is drunk, the ABI is isValid
+      abi: uniV2LPContract.abi,
+      function: 'approve',
```
Comment (gomesalexandre) [inline src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts:252]:
```suggestion
      functionName: 'approve',
```

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts [hunk 9]
```diff
@@ -234,10 +246,12 @@ export const useFoxFarming = (
 
     if (!adapter) throw new Error(`no adapter available for ${ethAsset.chainId}`)
 
-    const data = uniV2LPContract.interface.encodeFunctionData('approve', [
-      contractAddress,
-      MaxUint256,
-    ])
+    const data = encodeFunctionData({
+      // @ts-ignore this is drunk, the ABI is isValid
```
Comment (gomesalexandre) [inline src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts:250]: revert

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts [hunk 10]
```diff
@@ -263,7 +277,11 @@ export const useFoxFarming = (
 
     if (!adapter) throw new Error(`no adapter available for ${ethAsset.chainId}`)
 
-    const data = foxFarmingContract.interface.encodeFunctionData('getReward')
+    const data = encodeFunctionData({
+      // @ts-ignore this is drunk, the ABI is isValid
```
Comment (gomesalexandre) [inline src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts:281]: revert

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts (hunk 3 above)
Comment (gomesalexandre) [reply src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts]: 77b4140a97

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts (hunk 5 above)
Comment (gomesalexandre) [reply src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts]: 77b4140a97

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts (hunk 4 above)
Comment (gomesalexandre) [reply src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts]: 77b4140a97

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts (hunk 7 above)
Comment (gomesalexandre) [reply src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts]: 77b4140a97

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts (hunk 6 above)
Comment (gomesalexandre) [reply src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts]: 77b4140a97

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts (hunk 8 above)
Comment (gomesalexandre) [reply src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts]: 77b4140a97

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts (hunk 10 above)
Comment (gomesalexandre) [reply src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts]: 77b4140a97

File: src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts [hunk 11]
```diff
@@ -263,7 +277,11 @@ export const useFoxFarming = (
...
+      abi: foxFarmingContract.abi,
+      function: 'getReward',
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/features/defi/providers/fox-farming/hooks/useFoxFarming.ts]: 77b4140a97

File: src/contracts/abis/ERC20ABI.ts
Comment (gomesalexandre) [inline src/contracts/abis/ERC20ABI.ts:1]: Copied from wagmi - actually sane ERC20ABI which doesn't rug our viem contract instances

File: src/features/defi/providers/thorchain-savers/components/ThorchainSaversManager/Deposit/components/Approve.tsx [hunk 12]
```diff
@@ -194,13 +196,17 @@ export const Approve: React.FC<ApproveProps> = ({ accountId, onNext }) => {
...
-          amountCryptoBaseUnit,
-          quote.memo,
-          quote.expiry,
-        ])
+        const data = encodeFunctionData({
+          abi: thorContract.abi,
+          functionName: 'depositWithExpiry',
+          args: [
+            getAddress(quote.inbound_address),
+            getAddress(fromAssetId(assetId).assetReference),
```
Comment (0xApotheosis) [inline src/features/defi/providers/thorchain-savers/components/ThorchainSaversManager/Deposit/components/Approve.tsx:204]: That safety 🥰

**Review by 0xApotheosis — COMMENTED** [review]
Code is 🛍️ 

Will give a run-time check before stamping.

File: src/lib/viem-client.ts [hunk 13]
```diff
@@ -7,7 +7,7 @@ export const viemEthMainnetClient = createPublicClient({
   transport: http(getConfig().REACT_APP_ETHEREUM_NODE_URL),
 })
 
-export const viemBsclient = createPublicClient({
+export const viemBscClient = createPublicClient({
```
Comment (gomesalexandre) [inline src/lib/viem-client.ts:10]: 🧠  thanks @0xApotheosis !

File: src/contracts/contractManager.ts [hunk 14]
```diff
@@ -1,20 +1,19 @@
...
+import { THORChain_RouterABI } from 'contracts/abis/THORCHAIN_RouterABI'
 import { ethers } from 'ethers'
 import memoize from 'lodash/memoize'
+import type { Address } from 'viem'
+import { getContract } from 'viem'
 import { getEthersProvider } from 'lib/ethersProviderSingleton'
+import { viemEthMainnetClient } from 'lib/viem-client'
 
-import type { IUniswapV2Pair } from './__generated'
+import { erc20ABI } from './abis/ERC20ABI'
```
Comment (kaladinlight) [inline src/contracts/contractManager.ts:16]: throw this import up alongside the rest of the `contracts/abis` imports?

Comment (0xApotheosis) [reply src/contracts/contractManager.ts]: https://github.com/shapeshift/web/pull/5532/commits/b50bb9ddcfa4ef403a906deb978dd41e91732c7d

**Review by 0xApotheosis — APPROVED** [review]
Runtime sanity check looking good.

**Pull request merged**